tests/e2e/json_writer.go

```diff
@@ -0,0 +1,30 @@
+package main
+
+import (
+	"encoding/json"
+	"os"
+)
+
+// TraceWriter is an interface for writers that write steps to files.
+type TraceWriter interface {
+	WriteTraceToFile(filepath string, trace []Step) error
+}
+
+// JSONWriter is a simple writer that simply marshals the array of Step objects.
+// To identify which type of action is being used, we add a field to the Step struct.
+type JSONWriter struct{}
+
+var GlobalJSONWriter = JSONWriter{}
+
+func (writer JSONWriter) WriteTraceToFile(filepath string, trace []Step) error {
+	// collect missing action types, if any. this way, we can provide a more helpful error message.
+
+	// workaround: we would keep a set, but go doesn't have sets.
```

Comment:
I don't fully grasp this comment. This would be a workaround for what?
Could you use `map[X]bool` as a set?

Reply:
old comment that I did not remove, my bad! 55d268659